pkg/pillar/cmd/agent.go

```diff
@@ -0,0 +1,64 @@
+package cmd
+
+import (
+	"flag"
+	"github.com/lf-edge/eve/pkg/pillar/agentlog"
+	"github.com/lf-edge/eve/pkg/pillar/pidfile"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// AgentBase - Interface for all agents to use the AgentBase.
+// Each agent is expected to have a Context that implements
+// this interface.
+type AgentBase interface {
+	AgentBaseContext() *AgentBaseContext
+	AddAgentSpecificCLIFlags()
+	ProcessAgentSpecificCLIFlags()
+}
+
+// AgentBaseCLIParams - stores all the common cli params
+type AgentBaseCLIParams struct {
+	DebugOverride bool
+}
+
+// AgentBaseContext - a struct that represents a general agent context
+type AgentBaseContext struct {
+	CLIParams   AgentBaseCLIParams
+	ErrorTime   time.Duration
+	WarningTime time.Duration
+	AgentName   string
+}
+
+// processCLIFlags - Add flags common to all agents
+func processCLIFlags(agentBase AgentBase) {
+	debugPtr := flag.Bool("d", false, "Debug flag")
+	agentBase.AddAgentSpecificCLIFlags()
+	flag.Parse()
+	ctx := agentBase.AgentBaseContext()
+	ctx.CLIParams.DebugOverride = *debugPtr
+	agentBase.ProcessAgentSpecificCLIFlags()
+}
+
+// Run - a general run function that will handle all of the common code for agents
+func Run(agentSpecificContext AgentBase) {
+	processCLIFlags(agentSpecificContext)
+	ctx := agentSpecificContext.AgentBaseContext()
+	debugOverride := ctx.CLIParams.DebugOverride
+	if debugOverride {
+		log.SetLevel(log.DebugLevel)
+	} else {
+		log.SetLevel(log.InfoLevel)
+	}
+	err := agentlog.Init(ctx.AgentName, "")
+	if err != nil {
+		log.Fatal(err)
+	}
+	if err := pidfile.CheckAndCreatePidfile(ctx.AgentName); err != nil {
+		log.Fatal(err)
+	}
+	log.Infof("Starting %s\n", ctx.AgentName)
+	// Run a periodic timer so we always update StillRunning
```

Comment:
This comment is incorrect. This code just does the first StillRunning notification. The periodic timer is in the agents code.

Note that I don't think all agents (yet) use StillRunning, to this might make it look odd when you do a ls -lt /var/run/*.touch

Reply:
I guess we need to set a variable in AgentBaseContext NeedWatchdog that all agents using Watchdog will set?